tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php

```diff
@@ -0,0 +1,199 @@
+        } catch (CartRuleConstraintException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @When I add a restriction for cart rule :cartRuleReference, which requires at least :quantity product(s) in cart matching one of these rules:
+     * @When I add a restriction for cart rule :cartRuleReference, which requires any quantity of product(s) in cart matching one of these rules:
+     * @When I add a restriction for cart rule :cartRuleReference, which requires at least :quantity product(s), but I provide empty list of rules
+     *
+     * @param string $cartRuleReference
+     * @param int $quantity
+     * @param TableNode|null $table
+     *
+     * @return void
+     */
+    public function addRestrictionRule(string $cartRuleReference, int $quantity = 0, ?TableNode $table = null): void
+    {
+        $rules = [];
+
+        try {
+            if ($table) {
+                foreach ($table->getColumnsHash() as $row) {
+                    $rules[] = new RestrictionRule($row['type'], $this->referencesToIds($row['references']));
+                }
+            }
+
+            $restrictionGroups = [];
+            if (isset($this->productRestrictionGroupsByReference[$cartRuleReference])) {
+                $restrictionGroups = $this->productRestrictionGroupsByReference[$cartRuleReference];
+            }
+
+            $restrictionGroups[] = new RestrictionRuleGroup($quantity, $rules);
+            $this->productRestrictionGroupsByReference[$cartRuleReference] = $restrictionGroups;
+        } catch (CartRuleConstraintException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @When I restrict cart rules for :cartRuleReference providing non-existing cart rules
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function restrictCartRulesProvidingNonExistingIds(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+        $command->setRestrictedCartRuleIds([self::NON_EXISTING_CART_RULE_ID]);
+
+        try {
+            $this->getCommandBus()->handle($command);
+        } catch (CartRuleNotFoundException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @When I clear all product restrictions for cart rule :cartRuleReference
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function clearProductRestrictionRules(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+        $command->setProductRestrictionRuleGroups([]);
+
+        $this->getCommandBus()->handle($command);
+        $this->restrictionCommandsByReference[$cartRuleReference] = $command;
+        unset($this->productRestrictionGroupsByReference[$cartRuleReference]);
+    }
+
+    /**
+     * @When I save product restrictions for cart rule :cartRuleReference
```

Comment:
this "product" step can be removed. But if its the only thing to do then it shouldn't block this PR (I'll fix in upcoming PR)

Reply:
removed :heavy_check_mark: